Lint RST on GitHub Actions

hugovk opened the pull request:
<!--

Please include the PEP number in the pull request title, example:

PEP NNN: Summary of the changes made

In addition, please sign the CLA.

For more information, please read our Contributing Guidelines (CONTRIBUTING.rst)

-->

New version of https://github.com/python/peps/pull/1668, main difference is the linter has been updated so the `rst-backticks` skips code blocks (lines beginning with four spaces).

---

# Lint rules

This PR adds two reStructuredText lint rules and runs them on the CI using GitHub Actions. [Example run.](https://github.com/hugovk/peps/runs/1848734386?check_suite_focus=true)

1. Detect single backticks, which should be double in RST for inline code, e.g. stuff fixed in https://github.com/python/peps/pull/1554.

2. Detect inline code touching normal text, e.g. stuff fixed in https://github.com/python/peps/pull/1560.

# make lint

It adds a standalone `make` command to lint locally:

```sh
make lint
```

The pre-commit dependency is installed if needed.

# Optional git pre-commit hook

This uses the [pre-commit](https://pre-commit.com/) tool for linting. It's possible, but completely optional, to use it locally and it'll check your changes when you're committing, and fail before commit if something is found.

Optional local set up (run once):

```sh
pip install -U pre-commit && pre-commit install
```

Then make changes and commit as normal:

```sh
git commit -m "PEP XXX: etc"
```

Lint all files:

```sh
pre-commit run --all-files

# Or this does the same thing
make lint
```

The very first run takes a little while to set things up, after that it's quick.

But again, local use of pre-commit is completely optional for those who aren't keen, and the CI will run the checks anyway.

And `make lint` can be used completely independently of the git pre-commit hook.

gvanrossum reviewed (approved):
LGTM. Do you have powers to merge here?

FWIW I think there are a few more PEPs that fail this test, maybe you can re-trigger a run?

hugovk commented:
No, I don't have merge rights. I've rebased and fixed the new issues found PEP 646.

A couple of notes.

# Note 1

https://github.com/python/peps/pull/1802/commits/3d9f45cd634b2108fb09f9979f1f5dea7e2d0627

This:

```rst
(``Union`` is the one exception to this rule; see `Type Variable Tuples with ``Union```.)
```

Should be a link to this header:

```rst
Type Variable Tuples with ``Union``
-----------------------------------
```

But RST doesn't like code formatting inside links, `make pep-0646.html` gives:

```
docutils.utils.SystemMessage: pep-0646.rst:284: (ERROR/3) Unknown target name: "type variable tuples with ``union``".
```

See https://stackoverflow.com/q/4743845/724176 for ugly workarounds. So instead I've made the link without code formatting, and the HTML generation and link works.

# Note 2

https://github.com/python/peps/pull/1802/commits/9e1db9dd98a1f117311a26098ac3a0565e6fe416

@mrahtz:

This looks like it should be a link to a header:

```rst
Following `Type Variable Tuples Must Have Finite Length When Bound`, note
```

But there's no header with that name. "Type Variable Tuples Must Have Known Length" looks the closest to me. Is that the right one?

mrahtz commented:
Super nice work! I was always a bit worried about linting when writing PEP 646.

Note 2 - yes, that's right, it should be "Type Variable Tuples Must Have Known Length". Thanks!
